### ISSUE TITLE
[Training] GPU memory leakage for onnxruntime-training

### ISSUE THREAD

**Issue by weihChen** [microsoft/onnxruntime#17425]
### Describe the issue

When `pip install onnxruntime-training==1.15.1` and try to create the InferenceSession object repeatedly, the allocated GPU memory will not be released and end with `CUDA failure: out of memory` error.
However, if training ability is disabled and use `onnxruntime-gpu==1.15.1` instead, the above error disappear.

### To reproduce

```
import numpy as np
import onnx
import onnxruntime as ort
import torch
import torchvision


model_device = "cuda"
model_name = "resnet18"
model_pt = torchvision.models.__dict__[model_name](pretrained=True).to(model_device)

model_inputs = (torch.randn(1, 3, 224, 224, device=model_device),)
input_names = ["input"]
output_names = ["output"]
dynamic_axes = {"input": {0: "batch_size"}, "output": {0: "batch_size"}}
model_onnx_path = f"model_onnx/{model_name}.onnx"
torch.onnx.export(
    model_pt,
    model_inputs,
    model_onnx_path,
    input_names=input_names,
    output_names=output_names,
    opset_version=14,
    do_constant_folding=True,
    training=torch.onnx.TrainingMode.EVAL,
    dynamic_axes=dynamic_axes,
    export_params=True,
    keep_initializers_as_inputs=False,
)
model_onnx = onnx.load(model_onnx_path)
model_onnx = onnx.shape_inference.infer_shapes(model_onnx)

if __name__ == '__main__':
    num_model_trials = 500
    for model_trial_idx in range(num_model_trials):
        sess = ort.InferenceSession(model_onnx.SerializeToString(),
                                    providers=["CUDAExecutionProvider",
                                               "CPUExecutionProvider"])
        random_input = np.random.randn(128, 3, 224, 224).astype(np.float32)
        _ = sess.run(None, {"input": random_input})

```

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.15.1

### PyTorch Version

1.13.0+cu116

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.6

**Comment by skottmckay** [comment]
@askhade is this a known issue?

**Comment by askhade** [comment]
Interesting. This is not expected and I don't quite understand why this is a problem... the InferenceSession creation is not different in the Training Package. Needs more investigation. Will reply after root causing the issue.

**Comment by weihChen** [comment]
@askhade Hi, it seems not only InferenceSession. We have followed the official tutorial's instructions for [Train the Model on the Device](https://onnxruntime.ai/docs/api/python/on_device_training/training_api.html), GPU memory could not get released either.

**Comment by AdamLouly** [comment]
I can confirm that I was able to reproduce the problem with onnxruntime-training==1.15.1, but it appears to be resolved in version 1.16 (nightly). 

@askhade in these cases, how do we address the issue since its patched and 1.16 will be released soon?

**Comment by AdamLouly** [comment]
> @askhade Hi, it seems not only InferenceSession. We have followed the official tutorial's instructions for [Train the Model on the Device](https://onnxruntime.ai/docs/api/python/on_device_training/training_api.html), GPU memory could not get released either.

can you share how to reproduce it for this? I tried MNIST and could not reproduce it.

**Comment by weihChen** [comment]
@AdamLouly Thanks! With the nightly 1.6.0 version, the above oom issue of InferenceSession is resolved for me.

**Comment by weihChen** [comment]
@AdamLouly hi, I try the command `pip install -i https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/ORT-Nightly/pypi/simple/ onnxruntime-training ` but it reports the following error:
```
ERROR: Could not find a version that satisfies the requirement onnxruntime-training (from versions: none)
ERROR: No matching distribution found for onnxruntime-training
```
I can only find the nightly version without training ability from [ORT 1.16 Release Candidate Available for Testing](https://aiinfra.visualstudio.com/PublicPackages/_artifacts/feed/ORT-Nightly/PyPI/ort-nightly-gpu/overview/1.16.0.dev20230908001)

**Comment by baijumeswani** [comment]
To install nightly `onnxruntime-training`:

```
python -m pip install cerberus flatbuffers h5py numpy>=1.16.6 onnx packaging protobuf sympy setuptools>=41.4.0
pip install -i https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/ORT-Nightly/pypi/simple/ onnxruntime-training
```

**Comment by weihChen** [comment]
@baijumeswani Thanks! 
@AdamLouly My experiments show that the above OOM issue has been addressed with `onnxruntime-training==1.16.0`.
